pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+    Rebond,
+    Matching,
+    TransferToRelaychain,
+    RecordReward,
+    RecordSlash,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum ResponseStatus {
+    Ready,
+    Processing,
+    Successed,
+    Failed,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct Operation<BlockNumber> {
+    amount: Balance,
+    block_number: BlockNumber,
+    status: ResponseStatus,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct MatchingPoolBuffer {
+    total_unstake_amount: Balance,
+    total_stake_amount: Balance,
+    operation_type: Option<StakingOperationType>,
+}
+
+/// The single user's stake/unsatke amount in each era
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct MatchingUserBuffer {
+    /// The token amount that user unstake during this era, will be calculated
+    /// by exchangerate and xToken amount
+    total_unstake_amount: Balance,
+    /// The token amount that user stake during this era, this amounut is equal
+    /// to what the user input.
+    total_stake_amount: Balance,
+    /// The token amount that user have alreay claimed before the lock period,
+    /// this will happen because, in matching pool total_unstake_amount and
+    /// total_stake_amount can match each other
+    claimed_unstake_amount: Balance,
+    /// The token amount that user have alreay claimed before the lock period,
+    claimed_stake_amount: Balance,
+    /// To confirm that before lock period, user can only claim once because of
+    /// the matching.
+    claimed_matching: bool,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+
+    
+
+    use super::*;
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Currency type used for staking and liquid assets
+        type Currency: MultiCurrencyExtended<
+            Self::AccountId,
+            CurrencyId = CurrencyId,
+            Balance = Balance,
+            Amount = Amount,
+        >;
+
+        /// Currency used for staking
+        #[pallet::constant]
+        type StakingCurrency: Get<CurrencyId>;
+
+        /// Currency used for liquid voucher
+        #[pallet::constant]
+        type LiquidCurrency: Get<CurrencyId>;
+
+        /// The pallet id of liquid staking, keeps all the staking assets.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        /// Number of eras that user can get xToken back after stake on parachain.
+        #[pallet::constant]
+        type StakingDuration: Get<EraIndex>;
+
+        /// Number of eras that staked funds must remain bonded for.
+        #[pallet::constant]
+        type BondingDuration: Get<EraIndex>;
```

Comment:
this is a constant which means if relaychain changed this value, we also need to do runtime upgrade?

Reply:
I think when relaychain do runtime upgrade, parachain should do runtime upgrade too.